/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/flash -m 7 /data/result/result/Sample1/04.CutPrimer/Sample1_1.cutprimer.fq.gz /data/result/result/Sample1/04.CutPrimer/Sample1_2.cutprimer.fq.gz -o Sample1 -d /data/result/result/Sample1/04.CutPrimer && /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/bwa mem -Y -M -R "@RG\tID:Sample1\tSM:Sample1" -t 1 /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/database/nCov.fasta /data/result/result/Sample1/04.CutPrimer/Sample1.notCombined_1.fastq /data/result/result/Sample1/04.CutPrimer/Sample1.notCombined_2.fastq | /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/samtools view -b - | /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/samtools sort -T /data/result/result/Sample1/05.Stat/Sample1 -o /data/result/result/Sample1/05.Stat/Sample1.notCombined.bam - && /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/bwa mem -Y -M -R "@RG\tID:Sample1\tSM:Sample1" -t 1 /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/database/nCov.fasta /data/result/result/Sample1/04.CutPrimer/Sample1.extendedFrags.fastq | /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/samtools view -b - | /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/samtools sort -T /data/result/result/Sample1/05.Stat/Sample1 -o /data/result/result/Sample1/05.Stat/Sample1.extendedFrags.bam - && /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/samtools merge /data/result/result/Sample1/05.Stat/Sample1.bam /data/result/result/Sample1/05.Stat/Sample1.notCombined.bam /data/result/result/Sample1/05.Stat/Sample1.extendedFrags.bam && rm /data/result/result/Sample1/05.Stat/Sample1.notCombined.bam /data/result/result/Sample1/05.Stat/Sample1.extendedFrags.bam
/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/samtools index /data/result/result/Sample1/05.Stat/Sample1.bam
/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/mosdepth -n --fast-mode --by 100 /data/result/result/Sample1/05.Stat/depth /data/result/result/Sample1/05.Stat/Sample1.bam
zcat /data/result/result/Sample1/05.Stat/depth.regions.bed.gz|awk  '{print NR"\t"log($4+1)/log(10)}' > /data/result/result/Sample1/05.Stat/Sample1.draw.depth
/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/samtools depth -d 100000000 -a -b /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/database/nCoV.virus.cutprimer.bed /data/result/result/Sample1/05.Stat/Sample1.bam > /data/result/result/Sample1/05.Stat/Sample1.depth
export R_LIBS=/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/lib/R:$R_LIBS && /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/R-3.3.2/bin/Rscript /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/bin/line.depth.R /data/result/result/Sample1/05.Stat/Sample1.draw.depth /data/result/result/Sample1/05.Stat/Windows.Depth.svg
/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/freebayes -t /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/database/nCoV.variant.bed -H -p 1 -q 19 -m 60 --min-coverage 20 -F 0.6 -f /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/database/nCov.fasta /data/result/result/Sample1/05.Stat/Sample1.bam > /data/result/result/Sample1/05.Stat/Sample1.raw.vcf && /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/bcftools norm -f /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/database/nCov.fasta /data/result/result/Sample1/05.Stat/Sample1.raw.vcf -o /data/result/result/Sample1/05.Stat/Sample1.vcf
/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/bgzip -f /data/result/result/Sample1/05.Stat/Sample1.vcf
/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/tabix /data/result/result/Sample1/05.Stat/Sample1.vcf.gz
/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/jre1.8.0_101/bin/java -jar /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/bin/snpEff/snpEff.jar MN908947.3 /data/result/result/Sample1/05.Stat/Sample1.vcf.gz > /data/result/result/Sample1/05.Stat/Sample1.snpEff.vcf
export PYTHONPATH=/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/lib/python3-packages:$PYTHONPATH && /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/Python-3.7.5/bin/python3 /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/bin/get_anno_table.py /data/result/result/Sample1/05.Stat/Sample1.snpEff.vcf /data/result/result/Sample1/05.Stat Sample1
/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/bin/Consensus.pl /data/result/result/Sample1/05.Stat/Sample1.depth /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/database/nCov.fasta 30 /data/result/result/Sample1/05.Stat/Sample1.reference1.fa /data/result/result/Sample1/05.Stat/Sample1.vcf.gz
/data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/bcftools consensus -f /data/result/result/Sample1/05.Stat/Sample1.reference1.fa -o /data/result/result/Sample1/05.Stat/Sample1.Consensus.fa /data/result/result/Sample1/05.Stat/Sample1.vcf.gz
sed -i "s/MN908947.3 Wuhan seafood market pneumonia virus isolate Wuhan-Hu-1, complete genome/Sample1/g" /data/result/result/Sample1/05.Stat/Sample1.Consensus.fa
zcat /data/result/result/Sample1/05.Stat/Sample1.vcf.gz | grep -v '^#'|awk '{print $1"\t"$2-1"\t"$2"\t"$4"\t"$5}'| /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/tools/bedtools intersect -a - -b /data/pipeline/SARS-CoV-2_Multi-PCR_v1.1/database/wuhanRef.bed -loj |cut -f 1,3-5,9 > /data/result/result/Sample1/05.Stat/Sample1.vcf.anno
rm /data/result/result/Sample1/05.Stat/Sample1.draw.depth /data/result/result/Sample1/05.Stat/Sample1.reference1.fa
